### ISSUE TITLE
Mixing failsafehttp.RetryPolicy with hedgePolicy.HedgePolicy causes premature context cancel

### ISSUE THREAD

**Issue by yabberyabber** [failsafe-go/failsafe-go#65]
It looks like there's some interaction between failsafehttp.NewRoundTripper, hedgepolicy.HedgePolicy, and failsafehttp.RetryPolicy that causes the context associated with a successful request to be canceled prematurely. This makes it impossible to read the response body out of the response.

See the following minimal reproduction:
```go
func demoRequest() error {
        client := &http.Client{
                Transport: failsafehttp.NewRoundTripperWithExecutor(
                        http.DefaultTransport,
                        failsafe.NewExecutor[*http.Response](
                                failsafehttp.RetryPolicyBuilder().Build(),
                                // The hedge duration needs to be short enough that a hedged request
                                // is sent. If the response comes before the hedged request is sent
                                // then I do not observe the context cancellation error.
                                hedgepolicy.BuilderWithDelay[*http.Response](1*time.Nanosecond).Build(),
                        ),
                ),
        }
        req, err := http.NewRequest(http.MethodGet, "https://google.com/", nil)
        if err != nil {
                return err
        }
        resp, err := client.Do(req)
        // From my logging I can tell that the context associated with every request made
        // has already been canceled. I think the intention was to cancel all the outstanding
        // hedged requests but it appears that even the successful request is canceled.
        if err != nil {
                return err
        }
        defer resp.Body.Close()

        _, err = io.Copy(os.Stdout, resp.Body)
        if err != nil {
                // Cannot read resp.Body due to context cancellation error
                return err
        }
        return nil
}
```

Strangely I do not observe the context cancellations when I do not have a retrypolicy.  It seems to be some interaction of the two.

In this example reproduction I have set the hedge delay to 1 nanosecond which is absurdly small but it does force the race. I reality I have a hedge delay of a few seconds and occasional long-lasting requests.

My understanding is that the hedge policy will allow me to send alternate concurrent requests if the first request hangs for some reason; and that the retry policy will allow me to react to errors by sending another followup request. And my expectations was that by combining both policies, I would be able to send multiple concurrent requests and also handle some number of failures in earlier requests by retrying them. Perhaps there is a better way to accomplish this with the current API?

**Comment by jhalterman** [comment]
You're right - hedging doesn't play nice with http response consumption since all executions, including hedges, can share a single context which is canceled after any execution completes. It might be a fluke that you're not seeing this without a retry policy... I see it in either case.

I'll update the hedging internals shortly to avoid this.

**Comment by jhalterman** [comment]
I pushed a fix in 6de4fec9537c448f0fc0cde7ffc36d89d2d0f61e - let me know how that works for you and I can tag a new release with it. I wanted to get a failing test case with `httptest`, but couldn't get the context cancellation to be returned when reading the body as it does with http.Server, so I had to settle for hitting google.com for now.

**Comment by yabberyabber** [comment]
Thank you for the quick response, Jonathan --- this solution works for me and provides the expected behavior

**Comment by jhalterman** [comment]
This has been released in 0.6.8.
